src/test/java/org/jetbrains/plugins/ideavim/ex/implementation/services/VimVariableServiceTest.kt

```diff
@@ -113,4 +113,41 @@ class VimVariableServiceTest : VimTestCase() {
     typeText("n")
     assertState("1\n2\n1\n2\n1\n2\n1\n2\n${c}1\n")
   }
+
+  @Test
+  fun `test v hlsearch reflects highlight state`() {
+    configureByText(
+      """
+    |one two one
+    |${c}three four three
+    |five six five
+    """.trimMargin()
+    )
+    enterCommand("set hlsearch")
+
+    // Before any search, should be 0
+    enterCommand("echo v:hlsearch")
+    assertExOutput("0")
+
+    // After search with matches, should be 1
+    enterSearch("one")
+    enterCommand("echo v:hlsearch")
+    assertExOutput("1")
+
+    // After search with no matches, should be 0
+    enterSearch("xyz")
+    enterCommand("echo v:hlsearch")
+    assertExOutput("0")
+
+    // After nohlsearch command, should be 0 even though last search had matches
+    enterSearch("one")
+    enterCommand("nohlsearch")
+    enterCommand("echo v:hlsearch")
+    assertExOutput("0")
+
+    // After n command reusing previous search, should be 1 again
+    typeText("n")
+    enterCommand("echo v:hlsearch")
+    assertExOutput("1")
+  }
```

Comment:
As mentioned in the other PRs, extracting to a new file and splitting into separate functions would be good.

Reply:
I agree. I suggest we wait until the other PR for the `v:register` (#1046) variable is available so we can access the new `Variable` interface.